tools/autograd/templates/ViewFuncs.h

```diff
@@ -0,0 +1,149 @@
+#pragma once
+
+// ${generated_comment}
+
+#include <torch/library.h>
+
+#include <c10/core/SymIntArrayRef.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Operators.h>
+#else
+$ops_headers
+#endif
+
+namespace torch { namespace autograd {
+
+// forward declaration so it can be a friend class of ViewFunc
+struct ViewFuncSavedStateGuard;
+
+/// Base class for view functions, providing reapplication of a view on a new base.
+/// Each view op should get a codegenerated subclass of this class containing
+/// any state needed to reconstruct the view. The class also provides convenience
+/// accessors for saved SymInts / tensor state. This is useful for e.g. fake-ification,
+/// where we want to use symbolic values or fake tensors instead.
+struct TORCH_API ViewFunc {
+  virtual ~ViewFunc() {}
+  /// Returns any SymInts in the saved state.
+  virtual std::vector<c10::SymInt> get_symints() const = 0;
+  /// Returns any Tensors in the saved state.
+  virtual std::vector<at::Tensor> get_tensors() const = 0;
+  /// Reapplies the view on the given base using the saved state.
+  /// If specified, the provided values for SymInts / Tensors
+  /// are hot-swapped in for the view call.
+  virtual at::Tensor operator()(
+      const at::Tensor&,
+      const std::optional<std::vector<c10::SymInt>>& = c10::nullopt,
+      const std::optional<std::vector<at::Tensor>>& = c10::nullopt) const = 0;
+  // Allow limited setter access to maintain the const-ness invariant.
+  friend class ViewFuncSavedStateGuard;
+
+protected:
+  /// Sets the values of any SymInts in the saved state. The input vector size must
+  /// match the number of SymInts in the saved state (i.e. the size of the list
+  /// returned by get_symints()).
+  virtual void set_symints(const std::vector<c10::SymInt>&) = 0;
+  /// Sets the values of any Tensors in the saved state. The input vector size must
+  /// match the number of Tensors in the saved state (i.e. the size of the list
+  /// returned by get_tensors()).
+  virtual void set_tensors(const std::vector<at::Tensor>&) = 0;
+};
+
+// RAII guard for modifying and restoring saved SymInt / tensor state in a
```

Comment:
Hmm, I was not expecting to see this. In general, mutating a shared data structure is not thread safe. Because you are designing the class hierarchy here, it would not be too difficult to provide a virtual clone or clone_and_set method which gives you a new copy of the ViewFunc.

Reply:
Cool, I like the `clone_and_set()` idea - I'll go with that